src/impl_owned_array.rs

```diff
@@ -59,3 +69,558 @@ where
+            let tail_ptr = self.data.as_end_nonnull();
+            let mut tail_view = RawArrayViewMut::new(tail_ptr, array_shape, tail_strides);
+
+            if tail_view.ndim() > 1 {
+                sort_axes_in_default_order_tandem(&mut tail_view, &mut array);
+                debug_assert!(tail_view.is_standard_layout(),
+                              "not std layout dim: {:?}, strides: {:?}",
+                              tail_view.shape(), tail_view.strides());
+            } 
+
+            // Keep track of currently filled lenght of `self.data` and update it
+            // on scope exit (panic or loop finish).
+            struct SetLenOnDrop<'a, A: 'a> {
+                len: usize,
+                data: &'a mut OwnedRepr<A>,
+            }
+
+            impl<A> Drop for SetLenOnDrop<'_, A> {
+                fn drop(&mut self) {
+                    unsafe {
+                        self.data.set_len(self.len);
+                    }
+                }
+            }
+
+            let mut data_length_guard = SetLenOnDrop {
+                len: self.data.len(),
+                data: &mut self.data,
+            };
+
+
+            // Safety: tail_view is constructed to have the same shape as array
+            Zip::from(tail_view)
+                .and_unchecked(array)
+                .debug_assert_c_order()
+                .for_each(|to, from| {
+                    to.write(from.clone());
+                    data_length_guard.len += 1;
+                });
+            drop(data_length_guard);
+
+            // update array dimension
+            self.strides = strides;
+            self.dim = res_dim;
+        }
+        // multiple assertions after pointer & dimension update
+        debug_assert_eq!(self.data.len(), self.len());
+        debug_assert_eq!(self.len(), new_len);
+        debug_assert!(self.pointer_is_inbounds());
+
+        Ok(())
+    }
+}
+
+/// Sort axes to standard order, i.e Axis(0) has biggest stride and Axis(n - 1) least stride
+///
+/// The axes should have stride >= 0 before calling this method.
+fn sort_axes_in_default_order<S, D>(a: &mut ArrayBase<S, D>)
+where
+    S: RawData,
+    D: Dimension,
+{
+    if a.ndim() <= 1 {
+        return;
+    }
+    sort_axes1_impl(&mut a.dim, &mut a.strides);
+}
+
+/// Sort axes to standard order, i.e Axis(0) has biggest stride and Axis(n - 1) least stride
+///
+/// The axes should have stride >= 0 before calling this method.
+fn sort_axes_in_default_order_tandem<S, S2, D>(a: &mut ArrayBase<S, D>, b: &mut ArrayBase<S2, D>)
+where
+    S: RawData,
+    S2: RawData,
+    D: Dimension,
+{
+    if a.ndim() <= 1 {
+        return;
+    }
+    sort_axes_impl(&mut a.dim, &mut a.strides, &mut b.dim, &mut b.strides);
+}
+
+fn sort_axes1_impl<D>(adim: &mut D, astrides: &mut D)
+where
+    D: Dimension,
+{
+    debug_assert!(adim.ndim() > 1);
+    debug_assert_eq!(adim.ndim(), astrides.ndim());
+    // bubble sort axes
```

Comment:
Personally, I'd prefer an indirect sort with the standard library's sorting functions over a custom bubble sort.

Reply:
I have the feeling that the bubble sort is a good optimization for code size/perf for the common cases - 1-3 dimensional - but no evidence so far.

Reply:
I've already riced this to death in Dimension::_fastest_varying_stride_order, so this method should simply use that :)

Reply:
I don't think the performance of this function really matters much, at least for `append`, since it's called only once, and I'd expect copying the elements to dominate the execution time.

Reply:
I'm thinking that unfortunately append could be 5 elements or 5 million depending on the user. I don't mind keeping the simplest possible sort around - anything that sorts the dimensions in-place seems superior.

Reply:
As a note the benchmark `select_axis0` copies all of 256 elements of f32 but they are contiguous - that copying is so fast that this benchmark is quite sensitive to performance improvements in all the checking and dimension munging code around the copying, for example the improvement in 7564ad29109f7d8c77f875bb65aaac8ff7e7ffc2 was significant.